comb_spec_searcher/limited_strategy_db.py

```diff
@@ -0,0 +1,149 @@
+        """Search for a specification based on current data found."""
+        rules_dict = self.rules_up_to_equivalence()
+        # Prune all unverified labels (recursively)
+        # Afterward, only verified labels remain in rules_dict. In particular, there
+        # is a specification if <label> is in the rules_dict
+        if iterative:
+            rules_dict = iterative_prune(rules_dict, root=label)
+        else:
+            prune(rules_dict)  # this function removes rules not in a specification.
+
+        if self.mark_verified:
+            for ver_label in rules_dict.keys():
+                self.set_verified(ver_label)
+
+        # find the rules that we want to limit, and consider only those in a spec
+        rules_to_isolate = self.get_eqv_rules_using_strategies()
+        rules_to_isolate = set(
+            (lhs, rhs)
+            for (lhs, rhs) in rules_to_isolate
+            if lhs in rules_dict and rhs in rules_dict[lhs]
+        )
+
+        logger.debug("Found %s rules to isolate.", len(rules_to_isolate))
+
+        if len(rules_to_isolate) < self.limit:
+            rule_combos = [tuple(rules_to_isolate)]
+        else:
+            rule_combos = list(combinations(rules_to_isolate, self.limit))
+        num_combos = len(rule_combos)
+        logger.debug("%s combinations to check", num_combos)
+
+        # for each combination of <self.limit> rules, remove all but those and
+        # reprune to see if a spec still exists
+        for i, combo in enumerate(rule_combos):
+            logger.debug("[ %s / %s ]", i + 1, num_combos)
+            temp_rules_dict = deepcopy(rules_dict)
+            for lhs, rhs in rules_to_isolate:
+                if (lhs, rhs) in combo:
+                    continue
+                try:
+                    temp_rules_dict[lhs].remove(rhs)
+                except KeyError:
+                    pass
+                if len(temp_rules_dict[lhs]) == 0:
+                    del temp_rules_dict[lhs]
+            prune(temp_rules_dict)
+
+            if self.equivdb[label] in temp_rules_dict:
+                for label_in_spec in temp_rules_dict.keys():
+                    self.set_verified(label_in_spec)
+                if iterative:
+                    specification = iterative_proof_tree_finder(
+                        temp_rules_dict, root=self.equivdb[label]
+                    )
+                else:
+                    specification = self.smallish_random_proof_tree(
+                        temp_rules_dict, self.equivdb[label],
+                    )
+                return specification
+
+        raise SpecificationNotFound("No specification for label {}".format(label))
+
+    def smallish_random_proof_tree(self, rules_dict: RulesDict, root: int) -> Node:
+        """Searches a rule_dict known to contain at least one specification for a
+        small specification. Spends self.minimization_time_limit seconds searching."""
```

Comment:
should we look to add this method to the RuleDB so it can be used there too?

Reply:
I agree that it would be good to have that the RuleDBBase. It looks very similar to the feature proposed in #131. I think we should remove the function from this PR and open a separate one to add it to RuleDBBase (or maybe we're a happy with the way it's done in #131)